jujuclient/interface.go

```diff
@@ -231,6 +231,12 @@ type ControllerGetter interface {
 	// If there is no current controller, an error satisfying
 	// errors.IsNotFound will be returned.
 	CurrentController() (string, error)
+
+	// PreviousController returns the name of the previous controller used.
+	// If there is no previous controller (for instance controller has never been switched),
+	// an error satisfying errors.IsNotFound will be returned.
+	// It also retrieve if the last switch update the current controller or just shift the model
+	PreviousController() (string, bool, error)
```

Comment:
This signature is also a bad result of `HasControllerChangedOnPreviousSwitch`. This is quite nasty. It isn't at al obvious from the function name

If we can't improve the underlying data representation, I think this `bool` should be returned by it's own function

Reply:
I don't think it's a good things to add another method to the interface.
I don't see the "previous controller" and "the has Controller changed on previous switch" as independent data. They are coupled (they evolve in response to the same events).

However, if you don't like the return signature, maybe we can inverse the logic:

```go
	// If there is no previous controller (for instance controller has never been switched),
	// an error satisfying errors.IsNotFound will be returned.
	// if force is true, it will always return the previous controller.
	// If force is false, it may return the current controller if a previous switch hasn't changed the current controller.
	PreviousController(force bool) (string, error)
```

Reply:
I don't like this either really. Quite a conundrum

Reply:
Actually, thinking about it, this latest signature is worse, because it buries more logic in the store which it shouldn't (moreover, the store is completly mocked in the UT, so this later solution is worse, since the behavior won't be tested)

So, after sleeping on it we have two solution:

- Keep it that way (with a boolean returned by the PreviousController stuff)
- split into two method.

The former put the emphasis on the fact that "Previous controller" and "ControllerHasSwitched" is the same crap.

The later don't define such emphasis (which i think maybe an issue at some point), but may be easier to read.

Honestly, i can go for any tie break from @SimonRichardson, i don't have a strong enough conviction on it.